File: models/dto/permissions.go
package dto

import (
	"openauth/models/dao"
	"openauth/utils/customerrors"
)

type PermissionDetails struct {
	ID            int64  `json:"id"`
	Name          string `json:"name"`
	Description   string `json:"description"`
	CreatedByUser int64  `json:"createdByUser"`
	CreatedOn     int64  `json:"createdOn"`
	UpdatedOn     int64  `json:"updatedOn"`
}

type GetPermissionsResponse struct {
	Permissions []*PermissionDetails `json:"permissions"`
}

func (pd *PermissionDetails) FromPermission(perm *dao.Permission) *PermissionDetails {
	pd.ID = perm.ID
	pd.Name = perm.Name
	pd.Description = perm.Description
	pd.CreatedByUser = perm.CreatedByUser
	pd.CreatedOn = perm.CreatedOn
	pd.UpdatedOn = perm.UpdatedOn
	return pd
}

type PermissionDetailsShort struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}

type CreatePermissionRequest struct {
	Name          string `json:"name"`
	Description   string `json:"description"`
	UpdatedByUser int64  `json:"-"`
}

func (c *CreatePermissionRequest) Validate() error {
	if c.Name == "" {
		return customerrors.BAD_REQUEST_ERROR("name is required")
	}
	if c.Description == "" {
		return customerrors.BAD_REQUEST_ERROR("description is required")
	}
	return nil
}

type DeletePermissionRequest struct {
	PermissionId  int64 `json:"permId"`
	UpdatedByUser int64 `json:"-"`
}

func (c *DeletePermissionRequest) Validate() error {
	if c.PermissionId <= 0 {
		return customerrors.BAD_REQUEST_ERROR("invalid permission id")
	}
	return nil
}

type AddRemovePermissionsToUserRequest struct {
	UserId          int64   `json:"userId"`
	PermissionIDs   []int64 `json:"permIds"`
	UpdatedbyUserId int64   `json:"-"`
}

func (c *AddRemovePermissionsToUserRequest) Validate() error {
	if c.UserId <= 0 {
		return customerrors.BAD_REQUEST_ERROR("invalid userId")
	}
	if len(c.PermissionIDs) == 0 {
		return customerrors.BAD_REQUEST_ERROR("permissionIds are required")
	}
	return nil
}

type AddRemovePermissionsToGroupRequest struct {
	GroupId         int64   `json:"groupId"`
	PermissionIDs   []int64 `json:"permIds"`
	UpdatedbyUserId int64   `json:"-"`
}

func (c *AddRemovePermissionsToGroupRequest) Validate() error {
	if c.GroupId <= 0 {
		return customerrors.BAD_REQUEST_ERROR("invalid groupId")
	}
	if len(c.PermissionIDs) == 0 {
		return customerrors.BAD_REQUEST_ERROR("permissionIds are required")
	}
	return nil
}

File: admin/lib/models/permissions.dart
class PermissionDetails {
  int id;
  String name;
  String description;
  int createdByUser;
  int createdOn;
  int updatedOn;

  PermissionDetails({
    required this.id,
    required this.name,
    required this.description,
    required this.createdByUser,
    required this.createdOn,
    required this.updatedOn,
  });

  factory PermissionDetails.fromJson(Map<String, dynamic> json) {
    return PermissionDetails(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      createdByUser: json['createdByUser'],
      createdOn: json['createdOn'],
      updatedOn: json['updatedOn'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'createdByUser': createdByUser,
      'createdOn': createdOn,
      'updatedOn': updatedOn,
    };
  }
}

class GetPermissionsResponse {
  List<PermissionDetails> permissions;
  String error;

  GetPermissionsResponse({this.error = '', this.permissions = const []});

  factory GetPermissionsResponse.fromSuccessJson(Map<String, dynamic> json) {
    return GetPermissionsResponse(
      permissions: (json['permissions'] as List)
          .map((i) => PermissionDetails.fromJson(i))
          .toList(),
    );
  }
  factory GetPermissionsResponse.fromErrorJson(Map<String, dynamic> json) {
    return GetPermissionsResponse(
      permissions: json['error'],
    );
  }
}
